### PR TITLE
Improve feature configuration

### PR DESCRIPTION
**Pull request by pedromfedricci**
This PR intends to improve the feature selection for some common use cases. Closes: #271.
Users that would want to:
- 1.1 import http client only;
- 1.2 import ws client only;
- 2.1 import http server only;
- 2.2 import ws server only;
- 3.1 import http client and server only;
- 3.2 import ws client and server only;

The behaviour of the current feature flags is preserved, with one exception (clarification bellow), and adds more flags to provide for those above.
The exception: reexporting `RpcModule` and `SubscriptionSink` in the path [jsonrpsee::types::](https://github.com/paritytech/jsonrpsee/blob/5ae280a34c7205527ffbc8b4029b9be73c181099/jsonrpsee/src/lib.rs#L60) was removed as it is also reexported at root level [jsonrpsee::](https://github.com/paritytech/jsonrpsee/blob/5ae280a34c7205527ffbc8b4029b9be73c181099/jsonrpsee/src/lib.rs#L47) when server + macro are enabled.
This is to remove the path ambiguity, it also reduces the size of the resulting rlib for servers.
I didn't find the reason for having both paths, if there is I will revert it properly.

This table maps the described use cases against current flags and PR flags:
Note: compiled for x86_64-unknown-linux-gnu in debug mode.

| use case | current flag     | pr flag                             | rlib size reduction (%) |
| ------------ | ------------------ | -------------------------------- | ----------------------------- |
| 1.1          | client               | http-client                       | ~ 17                             |
| 1.2          | client               | ws-client                         | ~ 21                             |
| 2.1          | server             | http-server                      | ~ 20                             |
| 2.2          | server             | ws-server                       | ~ 26                             |
| 3.1          | client + server | http-client + http-server | ~ 11                             |
| 3.2          | client + server | ws-client + ws-server    | ~ 24                             |

Ty.

### CODE REVIEW

**Review by maciejhirsz — APPROVED** [review]
Looks good to me!

File: jsonrpsee/src/lib.rs [hunk 1]
```diff
@@ -24,38 +24,50 @@
...
-//! jsonrpsee wrapper crate.
+//! Jsonrpsee wrapper crate.
+//!
+//! <br>
+//!
+//! # Optional features
+//!
+//! Jsonrpsee crate composes JSON RPC functionality behind optional feature
+//! flags to provide for client and server communication over specific protocols.
+//! There are no default features, all functionality must be opted in accordingly.
```
Comment (dvdplm) [inline jsonrpsee/src/lib.rs:35]:
```suggestion
//! There are no default features, all functionality must be opted in to accordingly.
```

File: jsonrpsee/src/lib.rs [hunk 2]
```diff
@@ -24,38 +24,50 @@
...
 // DEALINGS IN THE SOFTWARE.
 
-//! jsonrpsee wrapper crate.
+//! Jsonrpsee wrapper crate.
+//!
+//! <br>
+//!
+//! # Optional features
+//!
+//! Jsonrpsee crate composes JSON RPC functionality behind optional feature
```
Comment (dvdplm) [inline jsonrpsee/src/lib.rs:33]:
```suggestion
//! The Jsonrpsee crate composes JSON-RPC functionality behind optional feature
```

File: jsonrpsee/src/lib.rs [hunk 3]
```diff
@@ -24,38 +24,50 @@
...
+#[cfg(feature = "jsonrpsee-ws-server")]
+pub use jsonrpsee_ws_server as ws_server;
 
 /// Set of RPC methods that can be mounted to the server.
-#[cfg(feature = "server")]
-pub use utils::server::rpc_module::{RpcModule, SubscriptionSink};
+#[cfg(any(feature = "http-server", feature = "ws-server"))]
+pub use jsonrpsee_utils::server::rpc_module::{RpcModule, SubscriptionSink};
 
 /// Procedural macros for JSON RPC implementations.
```
Comment (dvdplm) [inline jsonrpsee/src/lib.rs:67]:
```suggestion
/// Procedural macros for JSON-RPC implementations.
```

File: jsonrpsee/src/lib.rs [hunk 4]
```diff
@@ -24,38 +24,50 @@
...
-	pub use ::types::*;
-
-	/// Set of RPC methods that can be mounted to the server.
-	#[cfg(feature = "server")]
-	pub use utils::server::rpc_module::{RpcModule, SubscriptionSink};
-}
+#[cfg(feature = "jsonrpsee-proc-macros")]
+pub use jsonrpsee_proc_macros as proc_macros;
+
+/// Common types used to implement JSON RPC server and client
```
Comment (dvdplm) [inline jsonrpsee/src/lib.rs:71]:
```suggestion
/// Common types used to implement JSON-RPC server and client
```

**Review by dvdplm — APPROVED** [review]

**Comment by pedromfedricci** [comment]
hey! I fixed the comments. ty.

**Comment by dvdplm** [comment]
@pedromfedricci ty so much for this! Do you mind sorting out the merge conflicts? :)

**Comment by pedromfedricci** [comment]
oh ok! master changed since I pushed this, didn't notice.
This changes things a bit, would you take a look again? @dvdplm @maciejhirsz
ty!

**Comment by maciejhirsz** [comment]
Cheers, will merge as soon as CI is done.

File: jsonrpsee/src/lib.rs [hunk 5]
```diff
@@ -24,42 +24,54 @@
...
+/// Common types used to implement JSON-RPC server and client.
+#[cfg(feature = "jsonrpsee-types")]
+pub use jsonrpsee_types as types;
 
-/// Common types used to implement JSON RPC server and client.
-#[cfg(any(feature = "types", feature = "macros"))]
-pub mod types {
-	pub use ::types::*;
+/// JSON-RPC WebSocket client convinience macro to build params.
+#[cfg(feature = "ws-client")]
```
Comment (dvdplm) [inline jsonrpsee/src/lib.rs:72]: This isn't quite right I think, the `rpc_params` macro should be available for bot websocket and http clients.

Comment (pedromfedricci) [reply jsonrpsee/src/lib.rs:72]: I see, can do any for ws-client and http-client then? I assumed ws only based on the comment

Comment (pedromfedricci) [reply jsonrpsee/src/lib.rs:72]: solved: #498

Comment (dvdplm) [reply jsonrpsee/src/lib.rs:72]: Yeah my comment was wrong, my bad!

**Comment by dvdplm** [comment]
@pedromfedricci As a small thank you we wanted to give you a tip in DOTs, feel free to comment with your polkadot address as instructed by the bot. :)

**Pull request merged**